### DISCUSSION TITLE
Airflow Doesn't Register Changes to Codebase or Connections

### DISCUSSION THREAD

**Discussion by semibroiled** [apache/airflow#40820] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.9.1

### What happened?

Changs I make on my code doesn't show up in Airflow. It says I still need to rectify a change which is already done

Examples of Bugs

1) If I write `import pandas as pd` it says I wrote just `import `and need to finish writing the line

2) If I write `List[str]` it thinks I wrote `List[str`and need to close the bracket

3) For connections, even if I define them in UI it says I haven't defined them

### What you think should happen instead?

If I write the connection name and define it, it should get picked up by Airflow.

If I write changes in codebase, it should get reflected.

### How to reproduce

Please make any class in airflow, and call it in a functional script and call it in a DAG.

This is how my DAG is structured 

### Operating System

Sonoma 14.5

### Deployment

Docker-Compose

### Deployment details

just docker-compose I guess 

### Anything else?

Every time. In my branch 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
I converted it into discussion because it's not at all clear what you mean and what problem you have. Airlfow does not do any parsing or register codebase changes. It's usually a job of the IDE of your choice and you should make sure you configure it in the way to properly autocomplete changes, save files etc. But maybe this is a complete misunderstanding of how things works, so in order for someone to help you in the discussion, I suggest you provide more details and explain what is your environment and where you are doing what in much more detailed way.

**Reply by semibroiled** [reply]
Ah. I see. My choice of words were poor sorry. 

It's really a weird kind of error where one of two things happen

1) If Airflow says there is an error somewhere in my code/DAG, I can fix it, but it still doesn't show that it is corrected

2) Even if something isn't wrong in the code, and I write something new, for example, `import pandas as pd`, on airflow it will say I wrote `import pand`. To Illustrate another example, if I write out a Type Hint `List[str]`, it will say I wrote `List[str` and it's missing a closing bracket. 

I've found no way to remedy that. Curiously enough my other DAGs in other branches work just fine. 

Even if I do a docker-compose down / docker-compose up, delete all cache files, start over, the same happens unfailingly.

Curiously enough, a colleague who tried to help me fix this bumped into a separate issue. If I name a file 'abc.py' and rename it to '123.py', for example, it still shows cache or errors as if abc.py still existed. And 123.py doesnt register as having been created 

I hope that clarifies what I meant a little more. I'd love to add screenshot but I don't know how much that helps without the code itself

**Reply by semibroiled** [reply]
I don't know if this is something that contributes to this behaviour, but that is the only branch where I implement and call a DAG with a class and its methods. If I run as a script everything works fine, its just that airflow doesnt run it

**Reply by potiuk** [reply]
I read your second description several times and have no idea what is wrong and what is going on. I suggest you stop explaining with your own words, but copy & paste the snippets of your DAG and commands you run and errrors you see rather than trying to describe it - if you want someone to help you.

**Reply by semibroiled** [reply]
Hi @potiuk , I still don't see how it doesn't clarify things. I'll give it some more thought

Here's the Broken DAG error from Airflow

```
Broken DAG: [/opt/airflow/dags/xxx_internal/tempo/xxx_tempo_cost_tracking.py]
Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/opt/airflow/dags/xxx_internal/tempo/xxx_tempo_cost_tracking.py", line 7, in <module>
    from etl_scripts.tempo.xxx_tempo_scripting_functions import (
  File "/opt/airflow/plugins/etl_scripts/xxx/tempo/xxx_tempo_scripting_functions.py", line 93
    if month.isinstance() not (1 <= month <= 12):
                              ^
SyntaxError: invalid syntax
```

Here is the code it's referring to;
```python
if (isinstance(month, int)) and (not (1 <= month <= 12)):
        raise ValueError("Month must be between 1 and 12") 
```

Only errors get tracked and it gets stuck on it, and corrections don't actualise. Hope this illustrate what I mean further

**Reply by potiuk** [reply]
Can you please format it in the way that it will be readable?

**Reply by potiuk** [reply]
BTW. It looks like you have a syntax errror that you have to fix. But I am sorry to say it computers are much better in handling syntax errors so you should follow what the machine tells you

**Reply by potiuk** [reply]
`if month.isinstance() not (1 <= month <= 12):` is just not valid Python code you should fix it. But it's between you and Python interpreter to solve, there is no reason to involve othe people in your discussion with the interpreter.

Possibly you have problems with different versions of the interpreter, but again this is between you and your interpreter.

**Reply by semibroiled** [reply]
This is getting nowhere. Do you see any syntax error in the code itself? I just pasted the excerpt. My linters will also show no error in the code. Unit tests will pass. And I can run the function as a script .So why is airflow showing the code 3 hours in the past as it was being typed as an error and not showing the current state of the code which is correct? 

Do you still not understand? Or is it that you are unwilling to. I am genuinely at a loss.

**Reply by potiuk** [reply]
I have no idea how to help you. Python interpreter tells you that you are wrong. Now you are asking me to solve the problem for you . I am at a loss.

**Reply by semibroiled** [reply]
> `if month.isinstance() not (1 <= month <= 12):` is just not valid Python code you should fix it. But it's between you and Python interpreter to solve, there is no reason to involve othe people in your discussion with the interpreter.
> 
> Possibly you have problems with different versions of the interpreter, but again this is between you and your interpreter.

There is no such problem. The code is 

```
if (isinstance(month, int)) and (not (1 <= month <= 12)):
        raise ValueError("Month must be between 1 and 12")
```

My issue is airflow doesn't reflect it as such.  ```if month.isinstance() not (1 <= month <= 12):``` isn't valid code. I know that. isinstance() is not a method int types take and the boolean operators aren't set up correctly. I purposefully wrote it wrong, waited, corrected it, and waited, to show you that the Airflow DAG will still show the error prone version and not the actual correct code. 

If it is still not clear to you, I will stop the discussion and take it that there is no real vested interest to address this bug from the devs

**Reply by potiuk** [reply]
BTW. From the compiler's theory It's impossible to judge stack trace with single line to assess if there is a syntax error. You need more context - that your interpreter has. If you dont't trust you interpreter, you are at a complete loss. Airflow does not do parsing. Python interpreter is. If you get error that there is a syntax error, it quite likely is. So search for it,.

**Reply by potiuk** [reply]
> If it is still not clear to you, I will stop the discussion and take it that there is no real vested interest to address this bug from the devs

Genereally devs here are not to solve your basic problems. This forum is where people in their free time try to help others.  I tried. And failed to help you .Sorry for that if you expected your problem to be solved by someone spending their free time rather than watching olympics game opening in my free time.

**Reply by semibroiled** [reply]
I have no problem from the interpreter, because there is no wrong code written! Clearly between me writing code and airflow picking it up, something is going wrong, because Airflow REFUSES to show the actual state of the code and is hung up on an error, whether syntax or whatever, that has long since been patched in code. 

I have no such problems with other DAG runners. 

Let me show you another example; 


```
# Import Data Processing Libraries
import pandas as pd
```

This is my pandas import line

Let me purposefully break this

```
# Import Data Processing Libraries
import pand
```

Here is python showing somethings fishy

```
[{
	"resource": "/Users/xxx/xxx/xxx-de-data-platform/etl_scripts/xxx_internal/tempo/xxx_internal_tempo_scripting_functions.py",
	"owner": "_generated_diagnostic_collection_name_#7",
	"code": {
		"value": "reportMissingImports",
		"target": {
			"$mid": 1,
			"path": "/microsoft/pyright/blob/main/docs/configuration.md",
			"scheme": "https",
			"authority": "github.com",
			"fragment": "reportMissingImports"
		}
	},
	"severity": 4,
	"message": "Import \"pand\" could not be resolved",
	"source": "Pylance",
	"startLineNumber": 5,
	"startColumn": 8,
	"endLineNumber": 5,
	"endColumn": 12
}]
```

Fair enough


If I just refresh airflow:

```
Broken DAG: [/opt/airflow/dags/xxx_internal/tempo/xxx_internal_tempo_cost_tracking.py]
Traceback (most recent call last):
  File "/opt/airflow/dags/xxx_internal/tempo/xxx_internal_tempo_cost_tracking.py", line 7, in <module>
    from etl_scripts.xxx_internal.tempo.xxx_internal_tempo_scripting_functions import (
  File "/opt/airflow/plugins/etl_scripts/xxx_internal/tempo/xxx_internal_tempo_scripting_functions.py", line 5, in <module>
    import pand
ModuleNotFoundError: No module named 'pand'
```
Perfect. It's smart enough to know somethings wrong. Now I know somethings wrong too for sure. Let me go back and fix it:

```
# Import Data Processing Libraries
import pandas as pd
```

Sure enough, no more errors thrown by the interpreter in IDE

Let me refresh Airflow again:

```
Broken DAG: [/opt/airflow/dags/xxx_internal/tempo/xxx_internal_tempo_cost_tracking.py]
Traceback (most recent call last):
  File "/opt/airflow/dags/xxx_internal/tempo/xxx_internal_tempo_cost_tracking.py", line 7, in <module>
    from etl_scripts.xxx_internal.tempo.xxx_internal_tempo_scripting_functions import (
  File "/opt/airflow/plugins/etl_scripts/xxx_internal/tempo/xxx_internal_tempo_scripting_functions.py", line 5, in <module>
    import pand
ModuleNotFoundError: No module named 'pand'
```

Ah. I see it hasn't caught up. No matter, I will docker-compose down and up again,:

=> Still the same error on airflow

**Reply by semibroiled** [reply]
> > If it is still not clear to you, I will stop the discussion and take it that there is no real vested interest to address this bug from the devs
> 
> Genereally devs here are not to solve your basic problems. This forum is where people in their free time try to help others. I tried. And failed to help you .Sorry for that if you expected your problem to be solved by someone spending their free time rather than watching olympics game opening in my free time.

I am open to try to solve it too if I can be pointed in the right direction. I don't misvalue the time and effort you and you peers spend keeping the community alive and vibrant. Not only do I appreciate it, I respect you wholeheartedly. And it is out of that respect and desire to improve that I wanted this bug addressed as an issue in the first place. I formulated my words poorly once again, I apologise. I don't mean to say the community maintainers should be at beck and call for anything.

I am at a loss with your general attitude and refusal to acknowledge the problem as anything other than mistyped lines or syntax errors, when that is not the case here. If even illustrated excerpts aren't enough to drive the discussion forward by any veritable degree, I see no reason to further invest time trying to play a small part in bettering a product that is dismissive of that input.

**Reply by potiuk** [reply]
I have absolutely no idea how to help you, sorry - maybe someone else will be able to - but I am absolutely too dumb to understand the problems you have, sorry I give up.

**Reply by potiuk** [reply]
Maybe you will find someone smarter to help you

**Reply by potiuk** [reply]
Keep on trying
